feat: in page sticky nav

alisonjoseph opened the pull request:
Closes #385 

## Todo

- [x] set first item to active as default when you navigate to the page


#### Changelog

**New**

- page nav component

#### Testing / reviewing

- Test asset detail page nav at all breakpoints. Make sure sticky nav highlights when you scroll to an item and also when you click. 



https://user-images.githubusercontent.com/2753488/161967302-3b2808dc-2f66-45cb-84b6-8e0fa410b18f.mov

alisonjoseph committed:
feat: inpage nav styles

alisonjoseph committed:
wip: page nav

alisonjoseph committed:
wip: scroll sticky

alisonjoseph committed:
wip: active class sticky nav

alisonjoseph committed:
wip: active class sticky nav

alisonjoseph committed:
wip: wip

alisonjoseph committed:
Merge branch 'main' into nav-page-sticky-385

alisonjoseph committed:
Merge branch 'nav-page-sticky-385' of https://github.com/carbon-design-system/carbon-platform into nav-page-sticky-385

alisonjoseph committed:
wip: trying to get sticky nav to work

alisonjoseph committed:
wip: move sticky nav to asset detail page

francinelucca committed:
fix(web-app): sticky page nav: multiple fixes

alisonjoseph committed:
wip: cleanup

francinelucca committed:
Merge branch 'main' of github.com:carbon-design-system/carbon-platform into nav-page-sticky-385

francinelucca committed:
refactor(web-app): cleanup, decrease cogn. compl., use useEventListener

francinelucca committed:
fix(web-app): cleanup sticky page  nav

alisonjoseph committed:
wip: remove extra div

alisonjoseph committed:
fix: cleanup css

alisonjoseph committed:
fix: merge conflicts

francinelucca reviewed (dismissed):
still LGTM 🎉

alisonjoseph committed:
fix: merge conflicts

alisonjoseph committed:
Merge branch 'main' into nav-page-sticky-385

jharvey10 committed:
Merge branch 'main' into nav-page-sticky-385

alisonjoseph committed:
Merge branch 'main' into nav-page-sticky-385

mattrosno commented on services/web-app/components/page-nav/page-nav.js:
```diff
@@ -0,0 +1,152 @@
+  }, [handleHashChange])
+
+  const checkIfSectionIdIsActive = (id) => {
+    const section = document.getElementById(id)
```
Do we want to search the entire document for the `id` or just what's inside the `contentRef` with a `contentRef.current.querySelectorAll()`. I don't think it really matters? But that might be safer?

alisonjoseph replied:
I could not get it to work using contentRef here

mattrosno commented on services/web-app/components/page-nav/page-nav.js:
```diff
@@ -0,0 +1,152 @@
+    const sectionHeight = section.offsetHeight
+    const sectionBoundingClientRect = section.getBoundingClientRect()
+    const sectionTopDistance = sectionBoundingClientRect.top
+    // Space between top of screen and where we want section to be "Active"
```
```suggestion
    // Space between top of screen and where we want section to be "active"
```

mattrosno commented on services/web-app/components/page-nav/page-nav.js:
```diff
@@ -0,0 +1,152 @@
+      const sectionTopDistance = section.getBoundingClientRect().top
+      if (
+        // Setting active class when the top of the section is at the top
+        // of the scree nand the bottom of the section is visible
```
```suggestion
        // of the screen and the bottom of the section is visible
```

mattrosno reviewed (commented):
This is working great! I didn't check the spec, but are the nav links "At a glance" supposed to be left justified with the breadcrumb above?
<img width="1168" alt="image" src="https://user-images.githubusercontent.com/1691245/162007907-11d0f55d-23cf-4a6b-a154-b3f811924dde.png">

Awesome that we're handing the edge case of short viewport and the sticky nav no longer becoming sticky when the footer scrolls into view. Can we put some bottom padding or something on the side nav so there's some space between it and the footer?
<img width="1168" alt="image" src="https://user-images.githubusercontent.com/1691245/162008372-7e308fac-ed60-4457-b362-64230531b6fd.png">

If you go to a URL in a new tab like this, are we expecting the page to auto-scroll down to that heading? I'm seeing it stay at the top. http://localhost:3000/assets/carbon-charts/latest/alluvial#dependencies

mattrosno commented on services/web-app/components/page-nav/page-nav.module.scss:
```diff
@@ -0,0 +1,83 @@
+    }
+  }
+
+  // Sticky side nav stylese
```
```suggestion
  // Sticky side nav styles
```

mattrosno commented on services/web-app/components/page-nav/page-nav.module.scss:
```diff
@@ -0,0 +1,83 @@
+
+    display: flex;
+    width: 100%;
+    height: 32px;
```
```suggestion
    height: spacing.$spacing-07;
```

alisonjoseph commented on services/web-app/styles/_base.scss:
```diff
@@ -17,6 +18,16 @@ html {
   --scrollbar-width: 0px;
+
+  /* nav height + default top spacing */
+  --scroll-padding: #{calc(spacing.$spacing-09 + spacing.$spacing-06)};
```
@mattrosno my thinking here was that we'll need to update this based on whether pages have tabs or not, once we have mdx pages with anchor links across the site. So not using it now, but will need to set it dynamically in the future.

mattrosno replied:
Got it, makes sense. Do we have an open issue to fix page tabs? @kingtraceyj the page tabs should be sticky, right?
<img width="1116" alt="image" src="https://user-images.githubusercontent.com/1691245/162010946-4c76c28c-5ad8-46f0-a648-65a6b9053fb0.png">

kingtraceyj replied:
that's right, page tabs should be sticky when scrolling like the current component pages.

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.js

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.js

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.module.scss

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.module.scss

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.js

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

alisonjoseph committed:
Update services/web-app/components/page-nav/page-nav.module.scss

Co-authored-by: Matt Rosno <matt.rosno@gmail.com>

mattrosno reviewed (changes requested):
I found a bug. Using Chrome but didn't test others.

1. Go here http://localhost:3000/assets/carbon-charts
2. Step into a page like Alluvial
3. Scroll to the bottom of the page
4. Using browser back, it's impossible to get back to http://localhost:3000/assets/carbon-charts

jeanservaas commented:
@mattrosno just wanted to make sure that all banner headings ("Alluvial" here) are using the $fluid-display-01 token.

mattrosno commented:
> @mattrosno just wanted to make sure that all banner headings ("Alluvial" here) are using the $fluid-display-01 token.

@jeanservaas we're using this.

```scss
.title {
  margin-bottom: spacing.$spacing-07;
  color: theme.$text-primary;
  @include type.type-style('display-01', true);
  @include mixins.line-clamp(2);
  @include breakpoint.breakpoint-down(md) {
    @include type.type-style('quotation-02');

    font-family: type.font-family('sans');
  }
}
```

This PR does not change the page header title styling. Can you verify using https://next.carbondesignsystem.com/assets/carbon-charts/latest/alluvial?

kingtraceyj commented:
agreed with @mattrosno with the previous comments on the alignment and bottom padding. our bottom padding from those footer elements is 160px or `$spacing-13`

kingtraceyj reviewed (changes requested):
agreed with @mattrosno with the previous comments on the alignment and bottom padding. our bottom padding from those footer elements is 160px or `$spacing-13`

alisonjoseph committed:
fix: add bottom padding and update type token

alisonjoseph committed:
fix: left position text alignment

alisonjoseph commented:
@mattrosno @kingtraceyj @jeanservaas 
>This is working great! I didn't check the spec, but are the nav links "At a glance" supposed to be left justified with the breadcrumb above?

Good catch, fixed the alignment. 

----

> Can we put some bottom padding or something on the side nav so there's some space between it and the footer?

Added margin-bottom `$spacing-13 `

----

> Just wanted to make sure that all banner headings ("Alluvial" here) are using the $fluid-display-01 token.

confirmed correct tokens, and updated to use the new v11 naming `$fluid-display-01`  vs` $display-01 `

----
**TODO**
- [ ] If you go to a URL in a new tab like this, are we expecting the page to auto-scroll down to that heading? I'm seeing it stay at the top. http://localhost:3000/assets/carbon-charts/latest/alluvial#dependencies
- [ ] fix back button functionality

alisonjoseph committed:
fix: spacing

francinelucca committed:
fix(web-app): backbutton bug and race condition sticky nav

alisonjoseph committed:
fix: focus state styles

kingtraceyj commented:
one other thing @jeanservaas noticed is that sticky nav should be top aligned with text: 
<img width="987" alt="image" src="https://user-images.githubusercontent.com/20582179/162484430-f4810197-510d-4d19-be58-43201a306aa0.png">

alisonjoseph commented:
Since we don't have text yet, should I move the text up, or the dashboard down 😅  Right now the top spacing above the nav is `spacing-11`

<img width="770" alt="Screen Shot 2022-04-08 at 11 37 00 AM" src="https://user-images.githubusercontent.com/2753488/162484697-cd024be4-da15-42a1-9e9e-517345fa8565.png">

jeanservaas commented:
Hey @alisonjoseph it's ok that we don't have the top text yet... it will share the same top margin with whatever kicks off the page... in this case the dashboard.

kingtraceyj commented:
<img width="864" alt="image" src="https://user-images.githubusercontent.com/20582179/162486945-890bdc20-8ed7-47c6-9228-ad8da97151e1.png">

alisonjoseph committed:
fix: dashboard top spacing

alisonjoseph commented:
Updated the spacing @kingtraceyj 

One other question, should the first nav item be active when you land on the page? Right now its only active when you either click on the first nav item, or scroll down to it.

kingtraceyj commented:
@alisonjoseph let's start with the first item active if possible

jdharvey-ibm commented on services/web-app/components/page-nav/page-nav.js:
```diff
@@ -0,0 +1,163 @@
+    })
+  }
+
+  const linkClicked = (id) => {
```
minor: Some of the other event handlers are using `handle....()` as a naming convention. Could potentially use that here as well (i.e. `handleLinkClick`)

francinelucca replied:
Done!

francinelucca committed:
fix(web-app): start-off first item selected, review comments

francinelucca committed:
fix(web-app): rename click handler

francinelucca commented:
> @alisonjoseph let's start with the first item active if possible

Done! first element now class active by default

alisonjoseph committed:
fix: anchor link spacing

alisonjoseph committed:
fix: add label for a11y violation breadcrumb

kingtraceyj reviewed (changes requested):
looking good! i'm wondering if there is a way to add padding above the heading when a user a clicks anchor link so that the title aligns with the sticky nav. so instead of the title going to the very top of the page it stops 80px from the top. 

so instead of here:
![image](https://user-images.githubusercontent.com/20582179/162980422-b789f872-d41a-47f3-bb32-6587cef7c9e8.png)

it lands here:
![image](https://user-images.githubusercontent.com/20582179/162980655-92a93b2d-f564-4391-8d6a-899165aa51d8.png)

but once the correct padding is in place on real pages (for instance 80px under the dashboard) will it go ahead and do that?

alisonjoseph committed:
fix: update scrollpadding

kingtraceyj reviewed (dismissed):
looks good! 🚀

jdharvey-ibm commented on services/web-app/components/page-nav/page-nav.js:
```diff
@@ -93,11 +90,13 @@ const PageNav = ({ contentRef, items = [] }) => {
+    let activeSection = null
     // Space between top of screen and where we want section to be "Active"
-    const scrollDistance = 90
+    const scrollDistance = 120
```
Since this is used in more than one place in this file, it might be good to define it as a constant at the top.

alisonjoseph replied:
updated @jdharvey-ibm

alisonjoseph committed:
fix: scrolltopdistance const

francinelucca reviewed (approved):
😍

alisonjoseph committed:
Merge branch 'main' into nav-page-sticky-385

alisonjoseph committed:
Merge branch 'main' into nav-page-sticky-385

mattrosno reviewed (approved):
LGTM! I'll add an issue for us to populate actual anchor links from MDX in a later sprint.
